crates/bevy_reflect/src/path/access.rs

```diff
@@ -0,0 +1,240 @@
+    fn bad_enum(expected: Type, actual: impl Into<Type>) -> Self {
+        let actual = actual.into();
+        Error::Enum { expected, actual }
+    }
+    fn bad_type(expected: Type, actual: impl Into<Type>) -> Self {
+        let actual = actual.into();
+        Error::Type { expected, actual }
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+pub(super) enum Type {
+    Struct,
+    TupleStruct,
+    Tuple,
+    List,
+    Array,
+    Map,
+    Enum,
+    Value,
+    Unit,
+}
+
+impl fmt::Display for Type {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let name = match self {
+            Type::Struct => "struct",
+            Type::TupleStruct => "tuple struct",
+            Type::Tuple => "tuple",
+            Type::List => "list",
+            Type::Array => "array",
+            Type::Map => "map",
+            Type::Enum => "enum",
+            Type::Value => "value",
+            Type::Unit => "unit",
+        };
+        write!(f, "{name}")
+    }
+}
+impl<'a> From<ReflectRef<'a>> for Type {
+    fn from(value: ReflectRef<'a>) -> Self {
+        match value {
+            ReflectRef::Struct(_) => Type::Struct,
+            ReflectRef::TupleStruct(_) => Type::TupleStruct,
+            ReflectRef::Tuple(_) => Type::Tuple,
+            ReflectRef::List(_) => Type::List,
+            ReflectRef::Array(_) => Type::Array,
+            ReflectRef::Map(_) => Type::Map,
+            ReflectRef::Enum(_) => Type::Enum,
+            ReflectRef::Value(_) => Type::Value,
+        }
+    }
+}
+impl From<VariantType> for Type {
+    fn from(value: VariantType) -> Self {
+        match value {
+            VariantType::Struct => Type::Struct,
+            VariantType::Tuple => Type::Tuple,
+            VariantType::Unit => Type::Unit,
+        }
+    }
+}
+
+/// A singular owned element access within a path.
+///
+/// Can be applied to a `dyn Reflect` to get a reference to the targeted element.
+///
+/// A path is composed of multiple accesses in sequence.
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub(super) enum Access {
+    Field(Box<str>),
+    FieldIndex(usize),
+    TupleIndex(usize),
+    ListIndex(usize),
+}
+
+impl fmt::Display for Access {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Access::Field(field) => write!(f, ".{field}"),
+            Access::FieldIndex(index) => write!(f, "#{index}"),
+            Access::TupleIndex(index) => write!(f, ".{index}"),
+            Access::ListIndex(index) => write!(f, "[{index}]"),
+        }
+    }
+}
+
+/// A singular borrowed element access within a path.
+///
+/// Can be applied to a `dyn Reflect` to get a reference to the targeted element.
+///
+/// Does not own the backing store it's sourced from.
+/// For an owned version, you can convert one to an [`Access`] with [`AccessRef::to_owned`].
+#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+pub(super) enum AccessRef<'a> {
```

Comment:
Would it be possible to merge `AccessRef` into `Access` in a manner similar to `AssetPath`? Where `AccessRef` just becomes `Access<'a>` and `Access` becomes `Access<'static>`?

Reply:
yeah so I considered this, and didn't remember why I chose not to use a `Cow<'static, str>` here. And I think it's because then I can implement `Copy` on `AccessRef` and make the `element` methods take `self` over `&self`.

Though the `self` is not important all things considered and I'll take a look at how I can clean that up.